docs/source/features/memory_optimizations.rst

```diff
@@ -11,14 +11,16 @@ Flash Attention
 Overview
 ^^^^^^^^
 
-Flash Attention is a method designed to enhance the efficiency of Transformer models, which are widely utilized in applications such as Natural Language Processing (NLP). Traditional Transformers are slow and consume a lot of memory, especially with long sequences, due to the quadratic time and memory complexity of self-attention. FlashAttention, an IO-aware exact attention algorithm that leverages tiling to minimize the number of memory reads/writes between the GPU's high bandwidth memory (HBM) and on-chip SRAM. This approach is designed to be more efficient in terms of IO complexity compared to standard attention mechanisms.
+Flash attention is an algorithm designed to improve the efficiency of the attention mechanism in transformer models such as GPT and BERT. The attention mechanism has quadratic time and memory complexity in sequence length and can present significant runtime and memory challenges when training with long sequences. The flash attention algorithm, compared to standard attention, applies two techniques to lower the memory requirement and improve the compute efficiency. (i) Tiling: instead of working on the entire query, key, value tensors, it makes several passes at those input tensors and calculates the softmax one tile at a time. (ii) Recomputation: instead of storing the softmax matrix (qudratic to sequence length), it stores the softmax normalization factors (linear to sequence length) and recomputes attention scores using these factors in the backward pass. Despite an increased number of FLOPs due to recomputation, flash attention improves the runtime efficiency by minimizing the number of reads and writes between HBM and shared memory on the GPU. It also reduces the memory footprint from quadratic to linear to sequence length, greatly expanding the range of sequence length allowed for transformer models.
```

Comment:
Suggest a rewrite to break up the long paragraphs.

Flash attention is an algorithm designed to improve the efficiency of the attention mechanism in transformer models such as GPT and BERT. The attention mechanism has quadratic time and memory complexity in sequence length and can present significant runtime and memory challenges when training with long sequences. 

Compared to standard attention, the flash attention algorithm, applies two techniques to lower the memory requirement and improve the compute efficiency: tiling and recomputation.

The tiling technique makes several passes at those input tensors and calculates the softmax one tile at a time, instead of working on the entire query, key, and value tensors.

The recomputation technique stores the softmax normalization factors (linear to sequence length) and recomputes attention scores using these factors in the backward pass, instead of storing the softmax matrix (qudratic to sequence length).

Despite an increased number of FLOPs due to recomputation, flash attention improves the runtime efficiency by minimizing the number of reads and writes between HBM and shared memory on the GPU. It also reduces the memory footprint from quadratic to linear to sequence length, greatly expanding the range of sequence length allowed for transformer models.